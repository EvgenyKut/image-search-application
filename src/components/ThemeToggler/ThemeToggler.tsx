import React from 'react';
import { useAppSelector } from '../../hooks';
import useKeyListener from '../../hooks/useKeyListener';
import useLocalStorageState from '../../hooks/useLocalStorageState';
import { ReactComponent as Moon } from '../../assets/Moon.svg';
import { ReactComponent as Sun } from '../../assets/Sun.svg';
import styles from './themeToggler.module.css';

const ThemeToggler = () => {
  const { focus } = useAppSelector((state) => state.searchFocusReducer);
  const [theme, setTheme] = useLocalStorageState('search-app-theme', 'light');
  const nextTheme = theme === 'light' ? 'dark' : 'light';

  const setNextTheme = () => setTheme(nextTheme);
  useKeyListener('KeyF', setNextTheme, focus);

  React.useEffect(() => {
    document.body.dataset.theme = theme;
  }, [theme]);

  return (
    <div
      onClick={setNextTheme}
      className={styles.wrapper}
      data-testid="theme-toggler"
    >
      <div data-testid="dark" className={styles.darkTheme}>
        <Moon />
      </div>
      <div data-testid="light" className={styles.lightTheme}>
        <Sun />
      </div>
    </div>
  );
};

export default ThemeToggler;
