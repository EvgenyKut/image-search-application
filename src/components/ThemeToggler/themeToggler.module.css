.wrapper {
  display: flex;
  align-items: center;
  cursor: pointer;
  margin: 0 1rem;
}

.lightTheme {
  display: none;
}

.darkTheme {
  display: none;
}

body[data-theme='light'] .lightTheme {
  display: block;
}

body[data-theme='dark'] .darkTheme {
  display: block;
}
