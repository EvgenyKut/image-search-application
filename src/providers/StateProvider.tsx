import React, { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { setupStore } from '../store/store';

type StateProviderProps = {
  children: ReactElement;
};
const StateProvider: React.FC<StateProviderProps> = ({ children }) => {
  const store = setupStore();
  return <Provider store={store}>{children}</Provider>;
};

export default StateProvider;
