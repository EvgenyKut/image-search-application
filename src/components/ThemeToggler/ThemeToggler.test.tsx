import { render, screen } from '@testing-library/react';
import StateProvider from '../../providers/StateProvider';
import ThemeToggler from './index';
import renderer from 'react-test-renderer';

describe('ThemeToggler', () => {
  it('renders correctly ThemeToggler', () => {
    const tree = renderer
      .create(
        <StateProvider>
          <ThemeToggler />
        </StateProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should behave correctly', () => {
    render(
      <StateProvider>
        <ThemeToggler />
      </StateProvider>,
    );

    const toggler = screen.getByTestId('theme-toggler');
    expect(toggler).toBeInTheDocument();
    expect(toggler).toBeVisible();

    const lightMode = screen.getByTestId('light');
    expect(lightMode).toBeInTheDocument();
    expect(lightMode).toHaveStyle({ display: 'block' });
  });
});
