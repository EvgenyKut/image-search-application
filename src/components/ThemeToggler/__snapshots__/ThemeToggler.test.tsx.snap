// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeToggler renders correctly ThemeToggler 1`] = `
<div
  className="wrapper"
  data-testid="theme-toggler"
  onClick={[Function]}
>
  <div
    className="darkTheme"
    data-testid="dark"
  >
    <svg>
      Moon.svg
    </svg>
  </div>
  <div
    className="lightTheme"
    data-testid="light"
  >
    <svg>
      Sun.svg
    </svg>
  </div>
</div>
`;
